### PR TITLE
tests(backend):AIILG-439 templates tests

### PR DESCRIPTION
**Pull request by tlawal86**
# Overview
Add test coverage to the Templates module

## JIRA Ticket
https://mhclgdigital.atlassian.net/browse/AIILG-439

### CODE REVIEW

File: tests/backend/api/routes/test_templates.py [hunk 1]
```diff
@@ -1,13 +1,188 @@
...
+def mock_session():
+    session = AsyncMock()
+    session.exec = AsyncMock()
+    session.get = AsyncMock()
+    session.add = Mock()
+    session.commit = AsyncMock()
+    session.refresh = AsyncMock()
+    session.delete = AsyncMock()
+    return session
+
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py]: Similarly to here: https://github.com/communitiesuk/localai-local-transcribe/pull/121/changes#r3045128286, this looks like over-mocking

File: tests/backend/api/routes/test_templates.py [hunk 2]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
         response = await ac.get("/templates")
         assert response.status_code == expected_status_code
         assert len(response.json()) == len(TemplateManager.templates)
+
+
+@pytest.mark.asyncio
+async def test_get_templates_returns_list(mock_user):
+    res = get_templates(mock_user)
+    assert isinstance(res, list)
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py]:
```suggestion
```

I feel like we are already asserting this using mypy types no?

File: tests/backend/api/routes/test_templates.py [hunk 3]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
...
+
+@pytest.mark.asyncio
+async def test_get_user_template_not_found(mock_session, mock_user):
+    exec_result = Mock()
+    exec_result.first.return_value = None
+    mock_session.exec.return_value = exec_result
+
+    with pytest.raises(HTTPException) as exc_info:
+        await get_user_template(mock_user, mock_session, uuid.uuid4())
+    assert exc_info.value.status_code == 404
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:61]: Love this status code assert 🙌

File: tests/backend/api/routes/test_templates.py [hunk 4]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
...
+        content="Hello World",
+        description="test template",
+        type=TemplateType.DOCUMENT,
+        questions=[SimpleNamespace(position=1, title="Foo", description="foobar")],
+    )
+    monkeypatch.setattr("backend.api.routes.templates.UserTemplate", lambda **_: mock_user_template)
+
+    await create_user_template(mock_user, mock_session, request)
+    mock_session.add.assert_called_once_with(mock_user_template)
+    mock_session.commit.assert_awaited()
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py]: I like this test, I wonder if we could assert also that the created template has the right details?

Comment (Patryk-Kuchta) [reply tests/backend/api/routes/test_templates.py]: Could be a separate test or in this test if we rename it

File: tests/backend/api/routes/test_templates.py [hunk 5]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
...
+        name="New Request",
+        content=None,
+        description=None,
+        questions=[SimpleNamespace(position=1, title="Test Question", description="Hello World")],
+    )
+
+    await edit_user_template(mock_user, mock_session, mock_user_template.id, request)
+
+    mock_session.add.assert_called()
+    mock_session.commit.assert_awaited()
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:100]: Similar to above, I like this test, but I think we could be asserting more

File: tests/backend/api/routes/test_templates.py [hunk 6]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
...
+
+    mock_session.delete.assert_awaited()
+    mock_session.commit.assert_awaited()
+
+
+@pytest.mark.asyncio
+async def test_delete_user_template_not_found(mock_session, mock_user):
+    exec_result = Mock()
+    exec_result.first.return_value = None
+    mock_session.exec.return_value = exec_result
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:145]: info: is this part supposed to delete the template or I am just confused?

File: tests/backend/api/routes/test_templates.py [hunk 7]
```diff
@@ -11,3 +26,147 @@ async def test_get_templates_success(expected_status_code):
...
+@pytest.mark.asyncio
+async def test_duplicate_user_template_success(mock_session, mock_user, mock_user_template):
+    exec_result = Mock()
+    exec_result.first.return_value = mock_user_template
+    mock_session.exec.return_value = exec_result
+
+    await duplicate_user_template(mock_user, mock_session, mock_user_template.id)
+
+    mock_session.add.assert_called()
+    mock_session.commit.assert_awaited()
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:161]: assert to check that it's an actual duplicate?

File: tests/backend/api/routes/test_templates.py (hunk 5 above)
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:100]: I might be wrong about this but I think there a pottentially broken edge case when position=0 because of this line in the code:

        for question in request.questions:
            if isinstance(question, Question):
                existing_idx = next((i for i, q in enumerate(questions) if q.id == question.id), None)
                if existing_idx:
                    existing = questions.pop(existing_idx)
                    existing.title = question.title
                    existing.description = question.description
                    existing.position = question.position
                    continue

inside of edit user templates

File: tests/backend/api/routes/test_templates.py [hunk 8]
```diff
@@ -1,5 +1,20 @@
...
+    delete_user_template,
+    duplicate_user_template,
+    edit_user_template,
+    get_templates,
+    get_user_template,
+    get_user_templates,
+)
+from common.database.postgres_models import TemplateType
 from common.services.template_manager import TemplateManager
 from tests.utils import get_test_client
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:19]: I know it was already there but there parametrisation on line 22 is not great, could it be addressed please 🥺

Comment (Patryk-Kuchta) [reply tests/backend/api/routes/test_templates.py:19]: the asyncio market is inconsistent too (the one you added is better)

File: tests/backend/api/routes/test_templates.py
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_templates.py:1]: I think this is quite a big component with quite a lot of edge cases, especially we could expand the coverage of form template logic I think

**Review by Patryk-Kuchta — CHANGES_REQUESTED** [review]
Massive improvement in the style :D, a few comments to have a look through :)

File: tests/backend/api/routes/test_templates.py (hunk 4 above)
Comment (tlawal86) [reply tests/backend/api/routes/test_templates.py]: Added a new test to assert the right details were used during construction

File: tests/backend/api/routes/test_templates.py (hunk 5 above)
Comment (tlawal86) [reply tests/backend/api/routes/test_templates.py:100]: Added a few more assertions to test for expected behaviour

File: tests/backend/api/routes/test_templates.py (hunk 6 above)
Comment (tlawal86) [reply tests/backend/api/routes/test_templates.py:145]: Screenshot 2026-04-13 at 14 47 53

It's to ensure template resolves to None/falsy so as to trigger the exception

File: tests/backend/api/routes/test_templates.py (hunk 5 above)
Comment (tlawal86) [reply tests/backend/api/routes/test_templates.py:100]: You were right to be suspicious. As 0 resolves to falsy, questions at index 0 were being duplicated rather than edited. 
I made a change (would normally create a separate ticket but seemed apt to keep it here) to explicitly resolve the existing_idx check as not None. Included an accompanying test (test_edit_user_template_question_at_index_zero_is_updated) which failed with the previous code's state but now passes with this change as well as another (test_edit_user_template_adds_new_question_when_no_id) to validate that session.add works under expected circumstances

**Review by Patryk-Kuchta — APPROVED** [review]
:shipit:
